### PR TITLE
feat(client/api): validate ms users

### PR DESCRIPTION
**Pull request by moT01**
These are suggested changes from MS team.

Adds a way to link (and unlink) Microsoft usernames to freeCodeCamp accounts.
- adds a new collection to store the MS usernames
- adds two new endpoints, `POST /user/ms-username` (link) and `DELETE /user/ms-username` (unlink)

Adds a new view type for the trophy challenges, and a new way to submit them
- Once a user account is linked, they just click the verify trophy button and it checks the MS API to see if they've earned it.
- adds new endpoint `POST /ms-trophy-challenge-completed` to submit trophy challenges

And some other suggested fixes from MS team

TODO: Add all this to API project board.

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Comment by ShaunSHamilton** [comment]
Quick Question:
Are we sure we want to allow multiple Microsoft usernames per account?

File: curriculum/challenges/english/19-foundational-c-sharp-with-microsoft/write-your-first-code-using-c-sharp/perform-basic-operations-on-numbers-in-c-sharp.md [hunk 1]
```diff
@@ -18,7 +18,9 @@ This challenge will be partially completed on Microsoft's learn platform. Follow
 
 What is the value of the following result?
 
-`int result = 3 + 1 * 5 / 2;`
+```clike
```
Comment (IEvangelist) [inline curriculum/challenges/english/19-foundational-c-sharp-with-microsoft/write-your-first-code-using-c-sharp/perform-basic-operations-on-numbers-in-c-sharp.md:21]: What language slug is this?

File: api-server/src/server/boot/challenge.js [hunk 2]
```diff
@@ -121,6 +127,10 @@ const savableChallenges = getChallenges()
   .filter(challenge => challenge.challengeType === 14)
   .map(challenge => challenge.id);
 
+const msTrophyChallenges = getChallenges()
+  .filter(challenge => challenge.challengeType === 18)
+  .map(({ id, msTrophyId }) => ({ id, msTrophyId }));
```
Comment (IEvangelist) [inline api-server/src/server/boot/challenge.js:132]: Shouldn't you be importing the `challengeType` here?

```ts
import { challengeType } from '../../../../config/challenge-types.ts';

// omitted for brevity
```

Then you'd use the `const` value?

```suggestion
const msTrophyChallenges = getChallenges()
  .filter(challenge => challenge.challengeType === challengeType.msTrophy)
  .map(({ id, msTrophyId }) => ({ id, msTrophyId }));
```

File: api-server/src/server/boot/user.js [hunk 3]
```diff
@@ -93,8 +112,89 @@ function deleteUserTokenResponse(req, res) {
...
+      }
+
+      const { userName } = await msApiRes.json();
+
+      if (!userName) {
+        res.status(500);
+        throw new Error(
+          'An error occured trying to link your Microsoft account'
+        );
+      }
```
Comment (ShaunSHamilton) [inline api-server/src/server/boot/user.js:151]: Is this definitely a `5XX`? Could the username be _missing_ or `undefined` for another reason?

File: curriculum/challenges/english/19-foundational-c-sharp-with-microsoft/write-your-first-code-using-c-sharp/perform-basic-operations-on-numbers-in-c-sharp.md (hunk 1 above)
Comment (moT01) [reply curriculum/challenges/english/19-foundational-c-sharp-with-microsoft/write-your-first-code-using-c-sharp/perform-basic-operations-on-numbers-in-c-sharp.md:21]: From chatGPT:

"In Prism.js, the clike language heading is used to define a generic language category that encompasses a group of languages that share C-like syntax. This category includes programming languages that are structurally similar to the C programming language, such as C++, C#, Java, and many others."

That was the assumption I made, but I guess I wasn't sure until now (assuming GPT isn't lying). We already had it [included in our config](https://github.com/freeCodeCamp/freeCodeCamp/blob/d31e910f3a4078e01eec87eb181c544f32a0dea4/client/.babelrc.js#L50) so I just used that.

File: api-server/src/server/boot/challenge.js (hunk 2 above)
Comment (moT01) [reply api-server/src/server/boot/challenge.js]: Yup, added ✅ 

I will make an issue to fix the rest of the spots a magic number is used in that file.

File: api-server/src/server/boot/user.js (hunk 3 above)
Comment (moT01) [reply api-server/src/server/boot/user.js:151]: I suppose I can't be 100% sure, but it seems like it shouldn't ever be missing. Transcripts can only be created by users, and users have to have a username (I tried deleting mine, didn't work) - when you check the transcript, you get the username every time as far as I can tell. I don't know exactly what the API is doing, or what's stored in their database - so it's not impossible. This seems appropriate.

File: client/src/templates/Challenges/redux/completion-epic.js [hunk 4]
```diff
@@ -79,9 +84,13 @@ function postChallenge(update, username) {
         updateComplete(),
         submitChallengeComplete()
       ];
-      // TODO(Post-MVP): separate endpoint for trophy submission?
-      if (isTrophyMissing)
-        actions.push(createFlashMessage(trophyMissingMessage));
+
+      if (challengeType === challengeTypes.msTrophy && type === 'error') {
+        actions = [createFlashMessage(msTrophyError)];
```
Comment (Sboonny) [inline client/src/templates/Challenges/redux/completion-epic.js:89]: Submitting challenge will be disabled if the trophy fails to submit.

```suggestion
        actions = [createFlashMessage(msTrophyError), submitChallengeError()];
```

Comment (moT01) [reply client/src/templates/Challenges/redux/completion-epic.js:89]: nice catch @Sboonny 👍

**Comment by ShaunSHamilton** [comment]
Not this PRs fault, but I failed the exam, and still have the tick:
image

Refreshing the page removes the tick, but it should not be put there.

_Cannot claim certification, though_

File: client/src/redux/ms-username-saga.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      yield put(setIsProcessing(false));
+      yield put(createFlashMessage(message.linked));
+    } else {
+      yield put(setIsProcessing(false));
+      yield put(createFlashMessage(message.linkErr));
+    }
+  } catch {
+    yield put(setIsProcessing(false));
+    yield put(createFlashMessage(message.linkErr));
+  }
```
Comment (ShaunSHamilton) [inline client/src/redux/ms-username-saga.js:42]: I would not use another `isProcessing` flag. I would go for similar logic we use for `fetchUser`, `fetchUserComplete`, `fetchUserError`.

**Review by ShaunSHamilton — APPROVED** [review]
Aside from my slightly unrelated comment about the client showing a tick on exam fail, this looks good to me.

I do have one more knit-pick, but it is not a blocker for this PR:

**Review by Sboonny — APPROVED** [review]

**Comment by moT01** [comment]
Lets go with this then. I will look at the exam tick issue on the next PR.

**Pull request merged**